nd4j/nd4j-backends/nd4j-backend-impls/nd4j-cuda/src/main/java/org/nd4j/linalg/jcublas/buffer/BaseCudaDataBuffer.java

```diff
@@ -907,26 +928,179 @@ public void set(float[] data, long length, long srcOffset, long dstOffset) {
+    public void set(short[] data, long length, long srcOffset, long dstOffset) {
         switch (dataType()) {
             case BOOL:  {
                 val pointer = new BytePointer(ArrayUtil.toBytes(data));
                 copyDataFromSrc(pointer,length,offset,dstOffset);
 
             }
             break;
+            case UBYTE: {
+                data = ArrayUtil.cutBelowZero(data);
+            }
             case BYTE: {
                 val pointer = new BytePointer(ArrayUtil.toBytes(data));
                 copyDataFromSrc(pointer,length,offset,dstOffset);
 
             }
             break;
+
+            case SHORT: {
+                val pointer = new ShortPointer(data);
+                copyDataFromSrc(pointer,length,offset,dstOffset);
+
+            }
+            break;
+            case INT: {
+                val pointer = new IntPointer(ArrayUtil.toInts(data));
+                copyDataFromSrc(pointer,length,offset,dstOffset);
+
+            }
+            break;
+            case LONG: {
+                val pointer = new LongPointer(ArrayUtil.toLongs(data));
+                copyDataFromSrc(pointer,length,offset,dstOffset);
+
+            }
+            break;
+            case HALF: {
+                val pointer = new ShortPointer(ArrayUtil.toHalfs(data));
+                copyDataFromSrc(pointer,length,offset,dstOffset);
+
+            }
+            break;
+            case BFLOAT16: {
+                val pointer = new ShortPointer(ArrayUtil.toBfloats(data));
+                copyDataFromSrc(pointer,length,offset,dstOffset);
+
+            }
+            break;
+            case FLOAT: {
+                val pointer = new FloatPointer(ArrayUtil.toFloats(data));
+                copyDataFromSrc(pointer,length,offset,dstOffset);
+
+            }
+            break;
+            case DOUBLE: {
+                val pointer = new DoublePointer(ArrayUtil.toDoubleArray(data));
+                copyDataFromSrc(pointer,length,offset,dstOffset);
+
+            }
+            break;
+            default:
+                throw new UnsupportedOperationException("Unsupported data type: " + dataType());
+        }
+
+
+    }
+
+    /**
+     *
+     * PLEASE NOTE: length, srcOffset, dstOffset are considered numbers of elements, not byte offsets
+     *
+     * @param data
+     * @param length
+     * @param srcOffset
+     * @param dstOffset
+     */
+    public void set(byte[] data, long length, long srcOffset, long dstOffset) {
+        switch (dataType()) {
+            case BOOL:  {
+                val pointer = new BooleanPointer(ArrayUtil.toBooleanArray(data));
+                copyDataFromSrc(pointer,length,offset,dstOffset);
+
+            }
+            break;
+
             case UBYTE: {
-                for (int e = 0; e < data.length; e++) {
-                    put(e, data[e]);
-                }
+                data = ArrayUtil.cutBelowZero(data);
+            }
```

Comment:
Is this explicitly meant to fall through to the next case? If so, put a note here.